src/topology/sheaves/skyscraper.lean

```diff
@@ -0,0 +1,351 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import algebraic_geometry.sheafed_space
+import topology.sheaves.sheaf_condition.unique_gluing
+import topology.sheaves.stalks
+import category_theory.preadditive.injective
+import algebra.category.Group.abelian
+
+/-!
+# Skyscraper (pre)sheaves
+
+A skyscraper (pre)sheaf is a (pre)sheaf supported at a single point: if `p₀ ∈ X` is a specified
+point, then the skyscraper (pre)sheaf `𝓕` with value `A` is defined by `U ↦ A` if `p₀ ∈ U` and
+`U ↦ *` if `p₀ ∉ A` where `*` is some terminal object.
+
+## Main definitions
+
+* `skyscraper_presheaf`: `skyscraper_presheaf p₀ A` is the skyscraper presheaf at point `p₀` with
+  value `A`.
+* `skyscraper_sheaf`: the skyscraper presheaf satisfies the sheaf condition.
+
+## Main statements
+
+* `skyscraper_stalk_of_mem_closure₀`: if `y ∈ closure {p₀}` then the stalk of
+  `skyscraper_presheaf p₀ S` at `y` is `S`.
+* `skyscraper_stalk_of_not_mem_closure₀`: if `y ∉ closure {p₀}` then the stalk of
+  `skyscraper_presheaf p₀ S` at `y` is `*` the terminal object.
+
+-/
+
+noncomputable theory
+
+section
+
+open topological_space
+open category_theory category_theory.limits
+open Top
+open opposite
+
+universes u v w
+
+variables {X : Top.{u}} (p₀ : X) {C : Type v} [category.{w} C] (S : C)
+variables [has_terminal C] [Π (U : opens X), decidable (p₀ ∈ U)]
+
+/--
+A skyscraper presheaf is a presheaf supported at a single point: if `p₀ ∈ X` is a specified
+point, then the skyscraper presheaf `𝓕` with value `A` is defined by `U ↦ A` if `p₀ ∈ U` and
+`U ↦ *` if `p₀ ∉ A` where `*` is some terminal object.
+-/
+@[simps] def skyscraper_presheaf : presheaf C X :=
+{ obj := λ U, if p₀ ∈ unop U then S else terminal C,
+  map := λ U V i, if h : p₀ ∈ unop V
+    then eq_to_hom $ by erw [if_pos h, if_pos (le_of_hom i.unop h)]
+    else terminal.from _ ≫ eq_to_hom (if_neg h).symm,
+  map_id' := λ U,
+  begin
+    split_ifs,
+    { apply eq_to_hom_refl },
+    { simpa only [eq_comp_eq_to_hom] using terminal_is_terminal.hom_ext _ _, },
+  end,
+  map_comp' := λ U V W iVU iWV,
+  begin
+    by_cases hW : p₀ ∈ unop W,
+    { have hV : p₀ ∈ unop V := le_of_hom iWV.unop hW,
+      simp only [dif_pos hW, dif_pos hV, eq_to_hom_trans] },
+    { rw [dif_neg hW, eq_comp_eq_to_hom], apply terminal_is_terminal.hom_ext }
+  end }
+
+section
+
+variables {p₀}
+
+lemma skyscraper_presheaf_obj_of_mem {U : opens X} (h : p₀ ∈ U) :
+  (skyscraper_presheaf p₀ S).obj (op U) = S := if_pos h
+
+lemma skyscraper_presheaf_obj_of_not_mem {U : opens X} (h : p₀ ∉ U) :
+  (skyscraper_presheaf p₀ S).obj (op U) = terminal C := if_neg h
+
+end
+
+/--
+A skyscraper sheaf is a sheaf supported at a single point: if `p₀ ∈ X` is a specified
+point, then the skyscraper sheaf `𝓕` with value `A` is defined by `U ↦ A` if `p₀ ∈ U` and
+`U ↦ *` if `p₀ ∉ A` where `*` is some terminal object.
+-/
+def skyscraper_sheaf : sheaf C X :=
```

Comment:
I think the universe level for [`Top.presheaf.is_sheaf_iff_is_sheaf_opens_le_cover`](https://leanprover-community.github.io/mathlib_docs/topology/sheaves/sheaf_condition/opens_le_cover.html#Top.presheaf.is_sheaf_iff_is_sheaf_opens_le_cover) is too small

Reply:
Though, it might not be useful to keep `skyscraper_presheaf` universe level depending on three universe levels `u, v, w`, because to calculate stalks, universe level needs to be restricted anyway.

Reply:
`opens_le_cover` generalized at #16214.

Reply:
Thanks for the generalisation, by using `opens_le_cover`, I came up with this:
```lean
⟨_, (skyscraper_presheaf p₀ S).is_sheaf_iff_is_sheaf_opens_le_cover.mpr $ λ ι U, nonempty.intro
 { lift := λ c, if h : p₀ ∈ (presheaf.sheaf_condition.opens_le_cover_cocone U).X
    then c.π.app (op ⟨_, ⟨(opens.mem_supr.mp h).some, le_refl _⟩⟩) ≫ eq_to_hom
      begin
       dsimp, rw [if_pos h, if_pos (opens.mem_supr.mp h).some_spec],
      end
    else (terminal_is_terminal.if_neg h).from _,
   fac' := λ c j,
   begin
    dsimp, split_ifs with h0, swap,
    { exact (terminal_is_terminal.if_neg h0).hom_ext _ _, },
    by_cases h1 : p₀ ∈ (presheaf.sheaf_condition.opens_le_cover_cocone U).X;
    split_ifs, swap,
    { rw [eq_comp_eq_to_hom], exact (terminal_is_terminal.if_neg h1).hom_ext _ _, },
    rw [category.assoc, eq_to_hom_trans],
    transitivity c.π.app (op ⟨U (opens.mem_supr.mp h1).some ⊓ j.unop.obj, ⟨_, inf_le_left⟩⟩) ≫
      eq_to_hom _,
    rotate, { dsimp, rw [if_pos h0, if_pos], exact ⟨(opens.mem_supr.mp h1).some_spec, h0⟩ },
    { have := @nat_trans.naturality _ _ _ _ _ _ c.π j
        (op ⟨U (opens.mem_supr.mp h1).some ⊓ j.unop.obj, ⟨_, inf_le_left⟩⟩)
        (quiver.hom.op (hom_of_le _) : op j.unop ⟶ _), swap, exact inf_le_right,
      dsimp at this, simp only [category.id_comp] at this,
      have h : p₀ ∈ U _ ⊓ j.unop.obj := ⟨(opens.mem_supr.mp h1).some_spec, h0⟩,
      split_ifs at this with h', swap, { exact false.elim (h' h), },
      rw [this, category.assoc, eq_to_hom_trans, eq_to_hom_refl, category.comp_id] },
    { have := @nat_trans.naturality _ _ _ _ _ _ c.π
        (op ⟨_, ⟨(opens.mem_supr.mp h1).some, le_refl _⟩⟩)
        (op ⟨U (opens.mem_supr.mp h1).some ⊓ j.unop.obj, ⟨_, inf_le_left⟩⟩)
        (quiver.hom.op (hom_of_le inf_le_left)),
      dsimp at this, simp only [category.id_comp] at this,
      have h : p₀ ∈ U _ ⊓ j.unop.obj := ⟨(opens.mem_supr.mp h1).some_spec, h0⟩,
      split_ifs at this with h', swap, { exact false.elim (h' h), },
      rw [this, category.assoc, eq_to_hom_trans], },
   end,
   uniq' := λ s f h0,
   begin
    dsimp at *, split_ifs with h1, swap, { exact (terminal_is_terminal.if_neg h1).hom_ext _ _ },
    specialize h0 (op ⟨_, ⟨(opens.mem_supr.mp h1).some, le_refl _⟩⟩), 
    split_ifs at h0 with h2, swap, { exact false.elim (h2 (opens.mem_supr.mp h1).some_spec) },
    rw [←h0, category.assoc, eq_to_hom_trans, eq_to_hom_refl, category.comp_id],
   end }⟩
```

which is one line longer than before

Reply:
Thanks for trying out the approach! I just realized there is an easier approach since you've shown in the other PR that the skyscraper is the pushforward of a presheaf on the one-point space. I think you can show that a presheaf on a one-point space is a sheaf iff the empty set is mapped to a terminal object, and then use [Top.sheaf.pushforward_sheaf_of_sheaf](https://leanprover-community.github.io/mathlib_docs/topology/sheaves/functors.html#Top.sheaf.pushforward_sheaf_of_sheaf) and [Top.presheaf.is_sheaf_of_iso](https://leanprover-community.github.io/mathlib_docs/topology/sheaves/sheaf.html#Top.presheaf.is_sheaf_of_iso). Of course, the proof is only shorter when you already have the isomorphism, which isn't currently included in this PR but could be.

Reply:
By the way, about the approach you construct the skyscraper-stalk adjunction: you showed that the skyscraper is the pushforward from the one-point space, but didn't show that the stalk is the pullback to the one-point space. (Here "is" should be understood to mean a `nat_iso` between functors; I also implicitly identify a category C with terminal object and the category of sheaves valued in C on the one-point space; is this equivalence of categories in mathlib?) Once you realize stalk as a pullback, you are reduced to [Top.presheaf.pushforward_pullback_adjunction](https://leanprover-community.github.io/mathlib_docs/topology/sheaves/presheaf.html#Top.presheaf.pushforward_pullback_adjunction).

Reply:
I think if your goal is just the Godement resolution, then you can probably use the direct image definition directly, and don't need this explicit description of the skyscraper sheaf (but it's definitely nice to have).

Reply:
> By the way, about the approach you construct the skyscraper-stalk adjunction: you showed that the skyscraper is the pushforward from the one-point space, but didn't show that the stalk is the pullback to the one-point space. (Here "is" should be understood to mean a `nat_iso` between functors; I also implicitly identify a category C with terminal object and the category of sheaves valued in C on the one-point space; is this equivalence of categories in mathlib?) Once you realize stalk as a pullback, you can simply apply [Top.presheaf.pushforward_pullback_adjunction](https://leanprover-community.github.io/mathlib_docs/topology/sheaves/presheaf.html#Top.presheaf.pushforward_pullback_adjunction).

Interesting, I will try this.

Reply:
> I think if your goal is just the Godement resolution, then you can probably use the direct image definition directly, and don't need this explicit description of the skyscraper sheaf (but it's definitely nice to have).

I choose to define it as a product of skyscrapers main to prove enough injectiveness; because it is very easy to say when skyscrapers are injective; so given any sheaf F we embed it in product of skyscraper with value of stalks then embed each component into another skyscraper and then take product, we get an injective presentation (almost) for free

Reply:
Sorry for always delaying the response, I am currently distracted by something else

Reply:
No worries! It takes time to come up with the best approach. I agree it's easier to use the explicit description for the purpose of calculating stalks, and maybe also for the purpose of showing when the skyscraper is injective (I need to think about it).

And yes, I think it's reasonable to keep the explicit definition but prove the sheaf condition in a separate PR using isomorphism to pushforward of a sheaf on single point space.

In the meantime I've been able to golf down the two stalk isomorphisms significantly in [my branch](https://github.com/leanprover-community/mathlib/compare/jjaassoonn/sky_def...alreadydone:mathlib:jjaassoonn/sky_def?expand=1). The main lessons are: (1) you don't need to go through `stalk` in order to define the cocone and show the cocones are colimits; (2) you should use `c.w` instead of  `c.ι.naturality`.

Reply:
Upon looking at the other PR, it seems your proof of `skyscraper_sheaf_injective` already only depends on the stalk-skyscraper adjunction (and that the stalk functor preserves mono), and you can get to the adjunction more directly using the pushforward definition of skyscraper, so the explicit description is of no use there, if I understand correctly.

Reply:
Yes. I think you are right. So I think it makes sense to define Godement resolution in terms of skyscraper (no matter the explicit description or pushforward definition) so that
$$\mathcal{F} \to \prod_{x}\mathrm{Sky}(\mathcal{F}_x) \to \prod_x\mathcal I_x$$
can be easily shown to be a monomorphism into an Injective sheaf. ($\mathcal{I}_x$ comes from enough invectiveness and $\mathcal{F}_x$)

Reply:
Yeah, so let's figure out which definition is the easiest to work with to show monomorphism. I think you meant to write Sky(I_x) instead of I_x? It should be easy to show the pushforward of a mono is a mono, using the fact that a (pre)sheaf morphism is mono iff it's a mono on every open (this is stated in Stacks for Set-valued (pre)sheaves, but presumably holds more generally (also for functor categories)?).

As for the first map being mono, it's also enough to use the stalk-skyscraper adjunction. Two morphisms G→F agree after composing with the unit F→Sky_x(F_x) iff the corresponding morphisms G_x→F_x under the adjunction agree, so it suffices to use the fact that a sheaf morphism is determined by the induced morphisms on stalks. This doesn't work for presheaves though; I assume the original proof also doesn't?

An independent issue: I think the docstring here is problematic, because the skyscraper may have nonzero (non-terminal) stalks at points other than p0, so it may not be supported on a single point.

Reply:
> Yeah, so let's figure out which definition is the easiest to work with to show monomorphism. I think you meant to write Sky(I_x) instead of I_x? It should be easy to show the pushforward of a mono is a mono, using the fact that a (pre)sheaf morphism is mono iff it's a mono on every open (this is stated in Stacks for Set-valued (pre)sheaves, but presumably holds more generally (also for functor categories)?).

Sheaf morphism is mono iff it’s a mono on every open holds for every concrete category that preserves pullback, this is in the other PR containing the sky-stalk adjunct ion

Reply:
> An independent issue: I think the docstring here is problematic, because the skyscraper may have nonzero (non-terminal) stalks at points other than p0, so it may not be supported on a single point.

I thought `supported don a single point p` meant something like that `F(U)` is not zero if and only if `p \in U`. I have removed this phrase to clear confusion

Reply:
> Sheaf(C, X) morphism is mono iff it’s a mono on every open holds for every concrete category C that preserves pullback, this is in the other PR containing the sky-stalk adjunction. 

I actually didn't find the exact statement in that PR, but I think you may use [category_theory.nat_trans.mono_iff_mono_app](https://leanprover-community.github.io/mathlib_docs/category_theory/adjunction/evaluation.html#category_theory.nat_trans.mono_iff_mono_app) combined with [category_theory.Sheaf.hom.mono_iff_presheaf_mono](https://leanprover-community.github.io/mathlib_docs/category_theory/sites/sheafification.html#category_theory.Sheaf.hom.mono_iff_presheaf_mono) that you introduced earlier.